src/app.ts

```diff
@@ -237,28 +254,60 @@ async function clearAndReloadSpidStrategy(
   // tslint:disable-next-line: no-object-mutation
   app._router.stack = app._router.stack.filter(not(isLoginRoute));
 
-  await registerLoginRoute(app);
-  log.info("Spid strategy re-initialization complete.");
-  if (onRefresh) {
-    onRefresh();
+  // tslint:disable-next-line: no-let
+  let newSpidStratedy: passport.Strategy | undefined;
+  try {
+    // Inject a new SPID Strategy generate function in awilix container
+    container.register({
+      [SPID_STRATEGY]: awilix.asFunction(spidStrategy).singleton()
+    });
+    newSpidStratedy = await defaultModule.loadSpidStrategy();
+    log.info("Spid strategy re-initialization complete.");
+  } catch (err) {
+    log.error("Error on update spid strategy: %s", err);
+    log.info("Restore previous spid strategy configuration");
+    container.register({
+      [SPID_STRATEGY]: awilix
+        .asFunction(() => Promise.resolve(previousSpidStrategy))
+        .singleton()
+    });
+    throw new Error("Error while initializing SPID strategy");
+  } finally {
+    defaultModule.registerLoginRoute(
+      app,
+      newSpidStratedy ? newSpidStratedy : previousSpidStrategy
+    );
+    if (onRefresh) {
+      onRefresh();
+    }
   }
+  return newSpidStratedy ? newSpidStratedy : previousSpidStrategy;
 }
 
 /**
  * Sets an interval to reload SpidStrategy
  */
 export function startIdpMetadataUpdater(
   app: Express,
+  originalSpidStrategy: passport.Strategy,
```

Comment:
what's the difference between `previousSpidStrategy` and `originalSpidStrategy` ?

Reply:
`originalSpidStrategy` is the spidStrategy initialized at the first time on app startup. `newSpidStrategy` is the value updated or restored during update process. Only the first time when `startIdpMetadataUpdater` is called `newSpidStrategy` is undefined, and `originalSpidStrategy` is forward to `clearAndReloadSpidStrategy` method.